File: 20.ValidParentheses/20_ValidParentheses.py
#解法1 開放括號(Opening bracket)最後進 
# 但其對應的閉合括號(Closing bracket)要最先出才會是合法的括號
def isValid(s: str) -> bool:
    #字串長度
    strLength = len(s)
    #長度 == 1 不合法
    if strLength == 1:
        return False

    #放置開放括號  當遇到閉合括號 pop出與其比對是否符合
    stack = []
    #存放封閉括號對應的開放括號
    bracketDic = {
        ")": "(",
        "}": "{",
        "]": "["
    }

    #迴圈比對字串
    for bracket in s:
        #表示為開放括號
        if bracket not in bracketDic:
            stack.append(bracket)
        else:
            if not stack:
                #當有閉合括號但stack無值  表示不合法
                return False
            elif stack and stack.pop() != bracketDic[bracket]:
                #閉合括號時 stack pop 與當前閉合比對是否一致         
                return False

    #需判斷stack若還有值  表示不合法
    if stack:
        return False
    
    return True

File: 020.ValidParentheses/20_ValidParentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */

//解法1 開放括號(Opening bracket)最後進但是應該要最先出才會是合法的括號
const isValid = function (s) {
    if (s.length <= 1) return false;

    //1.遇到開放的括號  放進arr(stack)內 
    let stack = [];
    //存放封閉括號對應的開放括號
    let bracketMap = {
        ")": "(",
        "}": "{",
        "]": "["
    }

    /*跑迴圈 遇到開放括號就加入stack  
    遇到封閉的括號就對stack pop 來比對是否一致*/
    for (let i = 0; i < s.length; i++) {
        if (s[i] === "(" || s[i] === "{" || s[i] === "[") {
            stack.push(s[i]);
        } else {
            //對stack pop 並與s[i]比對是否一致
            if (stack.pop() !== bracketMap[s[i]]) return false;
        }
    }

    //需確認stack內是否還有值  若有 return false
    if (stack.length > 0) return false;

    return true;
};

//解法2

const isValid2 = function (s) {

    if (s.length == 0 || s.length % 2 != 0) {
        return false;
    }

    const closingParentheses = {
        "(": ")",
        "{": "}",
        "[": "]"
    }

    const stack = [];

    for (let i = 0; i < s.length; i++) {
        if (s[i] in closingParentheses) {
            stack.push(s[i]);
        } else if (s[i] != closingParentheses[stack.pop()]) {
            return false;
        }
    }

    return stack.length === 0;
}